map.go

```diff
@@ -102,26 +102,24 @@ func (ms *MapSpec) Copy() *MapSpec {
 	return &cpy
 }
 
-func (ms *MapSpec) clampPerfEventArraySize() error {
-	if ms.Type != PerfEventArray {
-		return nil
-	}
-
+func fixupPerfEventArraySize(maxEntries uint32) (uint32, error) {
 	n, err := internal.PossibleCPUs()
 	if err != nil {
-		return fmt.Errorf("perf event array: %w", err)
+		return 0, err
 	}
 
-	if n := uint32(n); ms.MaxEntries > n {
-		ms.MaxEntries = n
+	if n := uint32(n); maxEntries == 0 || maxEntries > n {
```

Comment:
Why aren't we fixing up all possible inputs? What would `maxEntries > 0 && maxEntries < n` mean if a user uses that size?

Reply:
Good point. When specifying a `maxEntries < PossibleCPUs()` (let's say, 1 on a multi-core system), only bpf progs scheduled to cpu 0 will be able to successfully obtain a ring buffer to emit events to. Executions on all other CPUs will fail to emit any events.

It's not clear to me what the path forward is here. Maybe we could start rejecting all non-zero values to perf array maxentries and publish it as a breaking change?

Reply:
The full context is here: https://github.com/cilium/ebpf/issues/209 TL;DR: some kernel selftests use a perf event array of size 1. Some users also do funky stuff.

> only bpf progs scheduled to cpu 0 will be able to successfully obtain a ring buffer to emit events to. Executions on all other CPUs will fail to emit any events.

This is because we set a CPU filter on perf event create? Maybe the linux selftest is just hopelessly buggy or maybe we're not doing the right thing.

> Maybe we could start rejecting all non-zero values to perf array maxentries and publish it as a breaking change? It would also mean we can remove the code dealing with the clamping and simplify the process a bit..

That's the old behaviour, which kept people creating issues.

Reply:
Cool, perhaps worth a comment on the if statement?

Reply:
Added a sentence which explains the < n case, let me know what you think

Reply:
Looks good, thanks. :+1: